### PR TITLE
major: New IPC system using lightweight Lambda actor queue

### PR DESCRIPTION
**Pull request by astubbs**
- [x] Documentation (if applicable)
- [x] Changelog
- [x] Review

### CODE REVIEW

**Review by astubbs — COMMENTED** [review]
couple of docs and a test

File: parallel-consumer-core/src/main/java/io/confluent/csid/actors/Actor.java [hunk 1]
```diff
@@ -0,0 +1,254 @@
...
+ * <p>
+ *
+ * @param <T> the Type of the Object to send closures to
+ * @author Antony Stubbs
+ */
+@Slf4j
+@ToString
+@EqualsAndHashCode
+@RequiredArgsConstructor
+// todo rename to ActorRef? ActorInterface? ActorAPI? Also clashes with field name.
```
Comment (astubbs) [inline parallel-consumer-core/src/main/java/io/confluent/csid/actors/Actor.java:72]: Name

File: parallel-consumer-core/src/main/java/io/confluent/csid/actors/Actor.java [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+public class Actor<T> implements IActor<T>, Interruptible {
+
+    private final T actorRef;
+
+    private volatile State state = State.ACCEPTING_MESSAGES;
```
Comment (astubbs) [inline parallel-consumer-core/src/main/java/io/confluent/csid/actors/Actor.java:77]: Change to atomic enum ref

File: parallel-consumer-core/src/main/java/io/confluent/csid/actors/Actor.java [hunk 3]
```diff
@@ -0,0 +1,254 @@
...
+
+    /**
+     * Single queueing point for all messages to the actor.
+     * <p>
+     * {@link LinkedBlockingDeque} is implemented as a simple doubly-linked list protected by a single lock and using
+     * conditions to manage blocking. Thread safe, highly performant, single lock.
+     */
+    @Getter(AccessLevel.PROTECTED)
```
Comment (astubbs) [inline parallel-consumer-core/src/main/java/io/confluent/csid/actors/Actor.java:85]: Private

File: parallel-consumer-core/src/main/java/io/confluent/csid/actors/Actor.java [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+    private final LinkedBlockingDeque<Runnable> actionMailbox = new LinkedBlockingDeque<>();
+
+    @Override
+    public void tell(final Consumer<T> action) {
```
Comment (astubbs) [inline parallel-consumer-core/src/main/java/io/confluent/csid/actors/Actor.java:89]: Docs

File: parallel-consumer-core/src/main/java/io/confluent/csid/actors/Actor.java [hunk 5]
```diff
@@ -0,0 +1,254 @@
...
+     * <p>
+     * Given the elements currently in the queue at the beginning of the method, processes them, but no more.
+     * <p>
+     * In other words - processes all elements currently in the queue, but not new ones which are added during
+     * processing (the execution of this method). We do this so that we finish predictably and have no chance of taking
+     * forever (which could happen if messages were continuously added between polling the queue for new messages).
+     * <p>
+     * Does not execute scheduled - todo remove scheduled to subclass?
+     */
+    // todo in interface?
```
Comment (astubbs) [inline parallel-consumer-core/src/main/java/io/confluent/csid/actors/Actor.java:158]: Delete

File: parallel-consumer-core/src/main/java/io/confluent/csid/actors/Actor.java [hunk 6]
```diff
@@ -0,0 +1,254 @@
...
+        process();
+    }
+
+    /**
+     * Perform the NO-OP interrupt.
+     * <p>
+     * Note: Might not have actually interrupted a sleeping {@link BlockingQueue#poll()} if there was also other work on
+     * the queue.
+     */
+    private void interruptInternalSync(Reason reason) {
```
Comment (astubbs) [inline parallel-consumer-core/src/main/java/io/confluent/csid/actors/Actor.java:246]: In-line

File: parallel-consumer-core/src/main/java/io/confluent/csid/actors/Actor.java [hunk 7]
```diff
@@ -0,0 +1,254 @@
...
+        log.debug("Interruption signal processed: {}", reason);
+    }
+
+    public enum State {
+        ACCEPTING_MESSAGES,
```
Comment (astubbs) [inline parallel-consumer-core/src/main/java/io/confluent/csid/actors/Actor.java:251]: Docs

File: parallel-consumer-core/src/main/java/io/confluent/csid/actors/IActor.java [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+import java.util.function.Consumer;
+import java.util.function.Function;
+
+/**
+ * todo docs
+ *
+ * @param <T>
+ * @author Antony Stubbs
+ */
+// todo remove?
```
Comment (astubbs) [inline parallel-consumer-core/src/main/java/io/confluent/csid/actors/IActor.java:18]: Decide - extract to pr

File: parallel-consumer-core/src/main/java/io/confluent/csid/actors/Interruptible.java [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+    @AllArgsConstructor
+    @RequiredArgsConstructor
+    @ToString
+    @EqualsAndHashCode
+    class Reason {
+//        /**
+//         * @deprecated should not be used, as bypasses the requirement of providing a reason
+//         */
+//        @Deprecated
+//        public static final Reason UNKNOWN = new Reason("unknown-reason");
```
Comment (astubbs) [inline parallel-consumer-core/src/main/java/io/confluent/csid/actors/Interruptible.java:34]: Delete

File: parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/state/PartitionStateManager.java [hunk 10]
```diff
@@ -416,4 +425,13 @@ private boolean isBlockingProgress(WorkContainer<?, ?> workContainer) {
         return workContainer.offset() < partitionState.getOffsetHighestSucceeded();
     }
 
+    public long getNumberOfAssignedPartitions() {
```
Comment (astubbs) [inline parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/state/PartitionStateManager.java:428]: Must merge upstream pr

File: parallel-consumer-core/src/test/java/io/confluent/csid/actors/ActorTest.java [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+        public String greet(String msg) {
+            return PREFIX + msg;
+        }
+    }
+
+    @Test
+    void tell() {
+        actor.tell(g -> g.setTold(MESSAGE));
+        actor.process();
+//        ManagedTruth.assertThat(greeter). // todo get TG working with Greeter class
```
Comment (astubbs) [inline parallel-consumer-core/src/test/java/io/confluent/csid/actors/ActorTest.java:47]: Move to branch

File: parallel-consumer-core/src/test/java/io/confluent/csid/actors/ActorTest.java [hunk 12]
```diff
@@ -0,0 +1,121 @@
...
+
+        // await AT LEAST delayMs before condition passes (earlier would mean there was no block during processing)
+        // trigger should fire after the delay, waking of the process block and processing the closure to set the value
+        Awaitility.await()
+                .atLeast(Duration.ofMillis(delayMs))
+                .untilAsserted(() -> assertThat(greeter.getTold()).isEqualTo(magic));
+
+        //
+        Truth.assertWithMessage("Thread returned from blocking process ok")
+                .that(returnedAfterBlocking.getCount()).isEqualTo(0);
```
Comment (astubbs) [inline parallel-consumer-core/src/test/java/io/confluent/csid/actors/ActorTest.java:119]: Countdown patch subject

File: parallel-consumer-core/pom.xml [hunk 13]
```diff
@@ -124,7 +125,11 @@
                     <useHas>false</useHas>
                     <useGetterForLegacyClasses>true</useGetterForLegacyClasses>
                     <releaseTarget>8</releaseTarget>
+                    <packages>
+                        <package>io.confluent.csid</package>
+                    </packages>
                     <classes>
+<!--                        <class>io.confluent.csid.actors.Greeter</class>-->
```
Comment (astubbs) [inline parallel-consumer-core/pom.xml:132]: Extract

**Review by astubbs — COMMENTED** [review]
...

File: parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/internal/AbstractParallelEoSStreamProcessor.java [hunk 14]
```diff
@@ -1250,49 +1174,37 @@ public void setLongPollTimeout(Duration ofMillis) {
...
 
     @Override
     public void pauseIfRunning() {
-        if (this.state == State.running) {
-            log.info("Transitioning parallel consumer to state paused.");
-            this.state = State.paused;
-        } else {
-            log.debug("Skipping transition of parallel consumer to state paused. Current state is {}.", this.state);
-        }
+        getMyActor().tellImmediately(me -> {
```
Comment (astubbs) [inline parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/internal/AbstractParallelEoSStreamProcessor.java:1188]: immediate?

File: parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/internal/DrainingCloseable.java [hunk 15]
```diff
@@ -23,50 +31,60 @@ enum DrainingMode {
...
 
     /**
-     * Close the consumer.
+     * Close the consumer, blocking until it has finished closing.
      *
      * @param timeout      how long to wait before giving up
      * @param drainingMode wait for messages already consumed from the broker to be processed before closing
      */
-    void close(Duration timeout, DrainingMode drainingMode);
+    // todo consider - do users want to have checked or unchecked exceptions when closing?
```
Comment (astubbs) [inline parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/internal/DrainingCloseable.java:86]: consider

**Review by astubbs — COMMENTED** [review]
....

File: parallel-consumer-core/pom.xml [hunk 16]
```diff
@@ -4,11 +4,12 @@
...
 
 -->
-<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
     <parent>
         <groupId>io.confluent.parallelconsumer</groupId>
         <artifactId>parallel-consumer-parent</artifactId>
-        <version>0.5.2.5-SNAPSHOT</version>
+        <version>0.6.0.0-SNAPSHOT</version>
```
Comment (astubbs) [inline parallel-consumer-core/pom.xml:12]: Must make a 6.x branch first

**Review by astubbs — COMMENTED** [review]
...

File: parallel-consumer-core/src/main/java/io/confluent/csid/actors/ActorImpl.java [hunk 17]
```diff
@@ -212,7 +212,8 @@ private void interruptInternalSync(Reason reason) {
         log.debug("Interruption signal processed: {}", reason);
     }
 
-    private enum ActorState {
+    // todo TG needs public - should be private
```
Comment (astubbs) [inline parallel-consumer-core/src/main/java/io/confluent/csid/actors/ActorImpl.java:215]: resolve

**Review by astubbs — COMMENTED** [review]
lgtm

File: parallel-consumer-core/src/main/java/io/confluent/csid/actors/Actor.java [hunk 18]
```diff
@@ -0,0 +1,144 @@
...
+     * @return true if the queue is empty
+     */
+    boolean isEmpty();
+
+    /**
+     * Start accepting messages.
+     * <p>
+     * Any messages sent before this will be rejected, as there's a chance they may never be processed.
+     */
+    // todo shouldn't this be triggerd through first process call?
```
Comment (astubbs) [inline parallel-consumer-core/src/main/java/io/confluent/csid/actors/Actor.java:136]: yes, resolve

File: parallel-consumer-core/src/main/java/io/confluent/csid/actors/ActorImpl.java [hunk 19]
```diff
@@ -0,0 +1,297 @@
...
+
+    @Override
+    public <R> Future<R> askImmediately(FunctionWithException<T, R> action) {
+        CompletableFuture<R> future = checkStateFuture(ActorState.ACCEPTING_MESSAGES);
+        var task = createRunnable(action, future);
+        getActionMailbox().addFirst(task);
+        return future;
+    }
+
+    // todo only used from one place which is deprecated
```
Comment (astubbs) [inline parallel-consumer-core/src/main/java/io/confluent/csid/actors/ActorImpl.java:180]: remove

File: parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/internal/AbstractParallelEoSStreamProcessor.java [hunk 20]
```diff
@@ -325,6 +291,12 @@ private void checkNotSubscribed(org.apache.kafka.clients.consumer.Consumer<K, V>
         }
     }
 
+    @Override
+    public void subscribe(String topic) {
+        subscribe(Collections.singletonList(topic));
+    }
+
+    // replace with facade delegate
```
Comment (astubbs) [inline parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/internal/AbstractParallelEoSStreamProcessor.java:299]:
```suggestion
    // replace with facade delegate - in controller refactor branch
```

File: parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/state/ProcessingShard.java [hunk 21]
```diff
@@ -75,12 +84,21 @@ public boolean isEmpty() {
         return entries.isEmpty();
     }
 
-    public long getCountOfWorkAwaitingSelection() {
-        return entries.values().stream()
-                // todo missing pm.isBlocked(topicPartition) ?
-                .filter(WorkContainer::isAvailableToTakeAsWork)
-                .count();
-    }
+    // not used
```
Comment (astubbs) [inline parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/state/ProcessingShard.java:87]: resolve

File: parallel-consumer-core/src/test/java/io/confluent/parallelconsumer/internal/ProducerManagerTest.java [hunk 22]
```diff
@@ -258,7 +258,7 @@ void producedRecordsCantBeInTransactionWithoutItsOffsetDirect() {
                 .commitMode(PERIODIC_TRANSACTIONAL_PRODUCER));
 
         try (var pc = module.pc()) {
-            pc.start();
+//            pc.start();
```
Comment (astubbs) [inline parallel-consumer-core/src/test/java/io/confluent/parallelconsumer/internal/ProducerManagerTest.java:261]: resolve if needed

**Comment by eddyv** [comment]
Closing - Stale.

**Pull request closed without merging**